# <img src="https://blogueaffaires.cogeco.ca/wp-content/uploads/2019/07/shutterstock_710262001-1920x1280-2.jpg" align=center style="width: 600px;">

# <span style="color:red">SAÉ INTRO CYBER - GIRAULT Adrien et ARVIN-BÉROD Maxence</span>

Bienvenue sur notre SAE INTRO CYBER qui sera sur la **Réalisation d'un CTF débutant orienté sur le réseau et sur l'utilisation de Python et Scapy**. </br>

###  <span style="color:red">Qu'est-ce qu'un CTF ?</span> 
<div class="alert alert-block alert-info bilan">
CTF signifie "Capture The Flag" .</br>
C'est une suite de challenges de cybersécurité à enchainer pour récupérer des informations que l'on appelle le "drapeau".
</div>
<img src="https://www.social-engineer.org/wp-content/uploads/2014/01/SECTF-WebGFX-1024x455.jpg" align=center style="width: 350px;"> 

***DEF CON***, *la plus grande conférence de hackers*

###  <span style="color:red">Le CTF à réaliser :</span> 


*Salut jeune Padawan, les Sith préparent une attaque contre une planète localisée dans la 
bordure médiane. Heureusement le maître Jedi Guillemin a intercepté la communication 
chiffrée envoyée par Dark Plageuis à Dark Sidious qui indique la planète qui sera détruite et la 
date de l’attaque. Il est parti en mission sur Tatouine et te confie la mission de décoder ce 
message. Fais vite énormément de vies en dépendent ! La communication a été enregistrée 
dans un fichier nommé McDiarmid.pcapng*

###  <span style="color:red">Pour ce CTF nous aurons besoins de :</span> 

- WireShark
- FileZilla
- Python
- Scapy

### <span style="color:red">WireShark : </span> 
`WireShark` est un **analyseur** de protocoles **réseau** libre et gratuit. </br>
Il est utilisé dans le dépannage et l’analyse de réseaux informatiques, le développement de protocoles, l’éducation et la rétro-ingénierie

<img src="https://www.nesabamedia.com/wp-content/uploads/2019/12/Wireshark.jpg" align=center style="width: 300px;">

### <span style="color:red">FileZilla : </span> 
`FileZilla` est un client **FTP, FTPS et SFTP,** développé sous la licence publique générale GNU. </br>
Il est intégré à la liste des logiciels libres préconisés par l’État français dans le cadre de la modernisation globale de ses **systèmes d’informations** (S.I.).

*Client : Mode de transaction entre plusieurs programmes et processus. L'un envoie des requêtes (le client), l'autre attend les requêtes des clients et y répond (le serveur).* 

*FTP : "File Transfert Protocole" est un protocole de communication destiné au partage de fichiers sur un réseau TCP/IP.*

*FTPS : "File Transfert Protocole Secure" est un protocole de communication destiné à l'échange informatique de fichiers sur un réseau TCP/IP, variante du FTP, sécurisé avec les protocoles SSL ou TLS.*



<img src="http://www.sanjaywebdesigner.com/articles/wp-content/uploads/2017/06/Antu_filezilla.svg_.png" align=center style="width: 250px;">

### <span style="color:red">Python : </span> 
`Python` est un langage de programmation. </br>
Il favorise la programmation structurée, fonctionnelle et orientée **objet** (la programmation orientée objet permet aux objets de communiquer entre eux).

- Il est doté d'un typage **dynamique fort** (il garantit que les types de données employés décrivent correctement les données manipulées)
- D'une **gestion automatique** de la mémoire par **ramasse-miettes** (sous-système informatique de gestion automatique de la mémoire) 
- D'un système de gestion d'exceptions (permet de gérer les conditions exceptionnelles pendant l'exécution du programme).



<img src="http://zencart-france.com/wp-content/uploads/2019/06/langage-de-programmation-python.jpg" align=center style="width: 350px;">

### <span style="color:red">Scapy : </span> 
`Scapy` est un logiciel **libre** de manipulation de **paquets réseau** écrit en `python`.

Il est capable d'**intercepter** le trafic sur un segment réseau, de **générer** des paquets dans un nombre important de protocoles, de réaliser une **prise** d'empreinte de la pile TCP/IP, de faire un **traceroute**, d'**analyser** le réseau informatique...



<img src="https://www.leixue.com/uploads/2020/08/Scapy.png" align=center style="width: 350px;">

<img src="http://www.theatre2lacte-lering.com/theatre2lacte/wp-content/uploads/2015/02/les-objectifs-800x200.png" align=center style="width: 500px;"> 


### <span style="color:darkgreen">Chapitre 1 : Analyse du protocole de transfert de fichier</span>

- Découverte de l'interception du transfert de fichier entre deux Siths sur WireShark

- Récupération de `login` et `password` d'un Sith

- Découverte de FileZilla


### <span style="color:darkgreen">Chapitre 2 : Installation de FileZilla</span>

- Configuration d'un serveur FileZilla

- Création d'un utilisateur

- Transfert de fichier entre un PC serveur et PC client


### <span style="color:darkgreen">Chapitre 3 : Récupération automatique du login et du mot de passe</span>

- Création d'un script permettant de récupérer en **dur** un login et mot de passe

- Création d'un script permettant de récupérer **automatiquement** un login et un mot de passe

### <span style="color:darkgreen">Chapitre 4 : Récupération du fichier transmis à partir du numéro de port TCP et décodage du message chiffrée</span>

- Création d'un programme qui récupère toutes les données transférées dans la capture Wireshark et les stocker dans un fichier nommé SW2.pdf

- Création d'un programme qui permet de déchiffrer le message du Sith.

### <span style="color:darkgreen">Chapitre 5 : Récupération automatique du numéro de port TCP négocié pour le transfert et du fichier</span>

- Création d'un script pour récupérer le numéro de port négocié pour le transfert de données ainsi que le nom du fichier à transférer codé en **dur**.

- Récupérer automatiquement le numéro de port négocié pour le transfert, le nom du fichier et le fichier.

### <span style="color:darkgreen">Chapitre 6 : Attaque MITM</span>

- a. Sniff les paquets FTP
- b. Récupère le login et mot de passe de l’utilisateur
- c. Identifie le numéro de port négocié pour un transfert de fichier ainsi que le nom du fichier et la fin de transfert
- d. Enregistrer le fichier avec le nom de fichier utilisé pour le transfert
- Qu'est-ce qu'une attaque MITM ?

# <span style="color:darkgreen">Chapitre 1 : Analyse du protocole de transfert de fichier</span>




### <span style="color:red">1) Ouverture du fichier McDiarmid.pcapng dans WireShark</span>

<img src="https://imgur.com/hpYfulU.png" align=center style="width: 800px;">

Nous voyons qu'il s'agit d'un `header` qui envoie une réponse ou il y'a du texte.


### <span style="color:red">2) Le protocole applicatif utilisée de cette communication</span>

<img src="https://imgur.com/hpYfulU.png" align=center style="width: 800px;">

L'envoi de ce fichier utilise le protcole applicatif **FTP**

Il a comme port **source** : 21 et port **destination** : 56230 </br>
**FTP** pour *File Transfert Protocol* est un protocole de communication destiné au partage de fichiers sur un réseau TCP/IP. </br>



### <span style="color:red">3) Adresse IP client et serveur</span>

<img src="https://imgur.com/hpYfulU.png" align=center style="width: 800px;">

- Adresse IP client : `192.168.35.24` 
- Adresse IP serveur : `192.168.35.235`

### <span style="color:red">4) Protocole de transport utilisé</span>


Il s'agit d'un protocole **TCP**, WireShark permet de ce repérer sur le modèle **OSI**. Nous pouvons aperçevoir de haut en bas la couche 1 (Physique) à la couche 5 (Sessions).


**Protocole TCP :**
<div class="alert alert-block alert-info bilan">
"Transmission Control Protocol" est un protocole de contrôle de transmissions permettant d'établir une connexion et de transmettre des données.</br>

- TCP permet de remettre en ordre les datagrammes en provenance du protocole IP
- TCP permet de vérifier le flot de données afin d'éviter une saturation du réseau
- TCP permet de formater les données en segments de longueur variable afin de les "remettre" au protocole IP
- TCP permet de faire circuler simultanément des informations provenant de sources (applications par exemple) distinctes sur une même ligne
- TCP permet enfin l'initialisation et la fin d'une communication de manière courtoise
</div>

<img src="http://electricala2z.com/wp-content/uploads/2017/10/osi-model.gif" align=center style="width: 300px;">

Le numéro de ports **TCP** utilisé pour le protocole applicatif **FTP** par défaut est **21**

### <span style="color:red">5) Message de bienvenue de ce site</span>


Lorsque nous nous connectons, le serveur nous renvoie le message : </br>

#########################################

    ##   Come over the dark side    ##

########################################

<img src="https://imgur.com/t7nyPXy.png" align=center style="width: 500px;">



### <span style="color:red">6) Login et mot de passe</span>


- Le **USER** est `Dark_Sidius`
- Le **PASS** est `Dark_Force`

<img src="https://imgur.com/AhTl5WU.png" align=center style="width: 600px;">

Le **USER** et le **PASS** sont transmis dans une *requête* provenant du client allant au *serveur*. Ensuite, le serveur *répond* qu'il faut entrer le **PASSWORD** et l'utilisateur transmet une *requête* avec le mot de passe écrit.

### <span style="color:red">7) IP et PORTS</span>

Le message utilisé pour préciser sur quelle adresse IP et quel numéro de port envoyer les données est : *Request: USER*

- IP source (client) : `192.168.35.24`
- IP destination (serveur) : `192.168.35.235`

On envoie les données sur l'adresse IP destination du `serveur` et le port destination du `serveur` : **21**

L'adresse IP et le numéro de de port sont indiqués sur la **couche 4** du modèle OSI pendant le **Transport**

<img src="https://imgur.com/BxkrNoj.png" align=center style="width: 600px;">


### <span style="color:red">8) Protocole passif ou actif ?</span>



**FTP Actif :**
<div class="alert alert-block alert-info bilan">
En mode actif, c'est le client FTP qui détermine le port de connexion à utiliser pour permettre le transfert des données. Port : 20</br>
    
• Le port de commande du client contacte le port de commande du serveur et lui donne son port de données.

• Le serveur donne un accusé de réception au port de commande du client.

• Le serveur établit une connexion entre son port de données et le port de données du client.

• Enfin, le client envoie un accusé de réception au serveur.

</div>    

**FTP Passif :**
<div class="alert alert-block alert-info bilan">
En mode passif, le serveur FTP détermine lui-même le port de connexion à utiliser pour permettre le transfert des données (data connexion) et le communique au client. Port : 21 </br>

• Le client contacte le port de commande du serveur et émet une commande PASV pour indiquer qu'il s'agit d'une connexion passive.

• Ensuite, le serveur donne son port de données d’écoute au client.

• Ensuite, le client établit une connexion de données entre le serveur et lui-même à l'aide du port indiqué. (le port est donné par le serveur)

• Enfin, le serveur envoie un accusé de réception au client.
    
</div>    

Le mode utilisé est le mode **FTP Passif** car le port est le 21 et le serveur FTP détermine lui-même le port de connexion à utiliser pour permettre le transfert des données.
    
<img src="https://imgur.com/XVOHvAw.png" align=center style="width: 300px;">




### <span style="color:red">9-10) Transfert de données et commandes</span>

Dans la capture, le transfert de données est représenté par le protocole applicatif `FTP-DATA`. </br>
Nous allons les filtrer pour en savoir le nombres :
    
<img src="https://imgur.com/RO1fzTv.png" align=center style="width: 600px;">

- Il y'a **90** transfert de données.

Elles sont initiées après ces trammes :

- FTP	Response: 150 Opening ASCII mode data connection for file list.

Et les données transférées sont des **octets** dont le fichier pdf se nomme *attaque_sith.pdf*

Ce fichier contient 128 718 **bytes** soit 128 718 **octets** soit 128,718 *kilooctets*

Il y'a : **87** paquets de 1460 bytes qui vont être transférés

Si les paquets envois que **maximum** 1460 bytes, c'est parce que la taille **maximale** MTU(Maximum Transmission Unit) de l'UDP que nous pouvons recevoir sans fragmentation est de **1460 bytes (octets)**.

### <span style="color:red">11) Fin du téléchargement</span>

Le message qui marque la fin du téléchargement est : </br>

*Response: 226 Transfer complete*

<img src="https://image.freepik.com/psd-gratuitement/jolie-petite-barre-de-progression-psd_28-560.jpg" align=center style="width: 300px;">



# <span style="color:darkgreen">Chapitre 2 : Installation et configuration de FileZilla</span>


### <span style="color:red">1-2) Configuration de FileZilla</span>


Nous nous connectons au serveur FileZilla et nous allons créer un utilisateur :
<img src="https://imgur.com/KUihIML.jpg" align=center style="width: 600px;">
<img src="https://imgur.com/8WNpNpj.jpg" align=center style="width: 400px;">

Notes : *Nous avons du enlever les pare-feu côter serveur pour que je puisse accéder au FTP de Adrien*

Ensuite, je peux  télécharger son fichier :

<img src="https://imgur.com/zJYchyj.jpg" align=center style="width: 400px;">

Voici les logs : 

<img src="https://imgur.com/bTGjaSH.jpg" align=center style="width: 400px;">


# <span style="color:darkgreen">Chapitre 3 : Récupération automatique du login et du mot de passe</span>


<img src="https://imgur.com/5LmSe0P.jpg" align=center style="width: 600px;">

Le `user` est **Dark_Sidius** et le `password`est **Dark_Force**

### <span style="color:red">2) Création des commandes pour récupérer le login et mdp de l'utilisateur</span>


In [9]:
from scapy.all import *

# rdpcap lit un fichier pcap ou pcapng (format Wireshark) et renvoie une liste de paquets
paquets=rdpcap("Wireshark/McDiarmid.pcapng")
print("##################################################")
print("Récupération de l'utilisateur et le mot de passe :")
print("##################################################\n")

# On récupère le 2ème paquets en données brutes indiqué par "RAW" (bytes notés b)
print("USER :", paquets[2][Raw].load.split(sep=None)) #affiche le contenu du paquet FTP ENCODÉ et séparer les chaines en fonction d'un séparateur (espace : sep = None)
print("USER :", paquets[2][Raw].load.split(sep=None)[1].decode('UTF8'))#affiche le contenu du paquet FTP DÉCODÉ
# On récupère le 4ème paquets en données brutes indiqué par "RAW" (bytes notés b)
print("PASSWORD :", paquets[4][Raw].load.split(sep=None))#affiche le contenu du paquet FTP ENCODÉ et séparer les chaines en fonction d'un séparateur (espace : sep = None)
print("PASSWORD :", paquets[4][Raw].load.split(sep=None)[1].decode('UTF8'))#affiche le contenu du paquet FTP DÉCODÉ

##################################################
Récupération de l'utilisateur et le mot de passe :
##################################################

USER : [b'USER', b'Dark_Sidius']
USER : Dark_Sidius
PASSWORD : [b'PASS', b'Dark_Force']
PASSWORD : Dark_Force


`Decode UTF8` nous permet de décoder notre format héxadécimal brute noté **b'**.

### <span style="color:red">3) Écrire un script pour récupérer automatiquement le login et mot de passe de l’utilisateur</span>


In [69]:
paquets=rdpcap("Wireshark/McDiarmid.pcapng")

for i in range(len(paquets)) : # parcours la capture wireshark :
    if paquets[i][Raw].load.split(sep=None)[0].decode('latin1') == "USER" : # si dans la première valeur de la liste paquet[i][Raw] = à USER :#
        user = paquets[i][Raw].load.split(sep=None)[1].decode('latin1') # prends la deuxième valeur de la liste paquet[i][Raw]#
    if paquets[i][Raw].load.split(sep=None)[0].decode('latin1') == "PASS" : #même chose pour le password#
        password = paquets[i][Raw].load.split(sep=None)[1].decode('latin1') #même chose pour le password#
print("le nom de l'utilisateur est : ",user) 
print("le mot de passe  est : ",password)

le nom de l'utilisateur est :  Dark_Sidius
le mot de passe  est :  Dark_Force


# <span style="color:darkgreen">Chapitre 4 : Récupération du fichier transmis à partir du numéro de port TCP et décodage du message chiffrée</span>


### <span style="color:red">1) Identifier le paquet transportant les données du fichier transféré</span>

Pour identifier le plus simplement le paquet transportant les données du fichier transféré dans WireShark, il suffit de trouver le **port** source.</br>
En loccurence, le port n° `20` (port lorsqu'il y'a des transferts de fichier sur le FTP).



### <span style="color:red">2) Créer un programme qui récupère toutes les données transférées dans la capture Wireshark et les stocker dans un fichier nommé SW2.pdf</span>

In [49]:
from scapy.all import *

#ouverture/création de mon fichier pdf SW2
fichier = open('SW2.pdf', 'ab')# a : ajouter du contenu et b : pour le format bytes

# rdpcap lit un fichier pcap ou pcapng (format Wireshark)
paquets=rdpcap("Wireshark/McDiarmid.pcapng")
print("##############################################################")
print("Récupération de toutes les données transférées en cours...")
print("##############################################################\n")


# pour i dans compris dans les paquets
for i in range(len(paquets)):
    if paquets[i][TCP].sport == 20: #si les paquets[i] de protocole TCP de port source 20
        fichier.write(bytes(paquets[i][Raw]))#alors ça écrit dans 'fichier' en bytes, les paquets du port source 20
fichier.close() #ferme 'fichier'

print("\n##########################")
print("Récupération terminé !")
print("##########################")


##############################################################
Récupération de toutes les données transférées en cours...
##############################################################


##########################
Récupération terminé !
##########################


### <span style="color:red">3) Ouverture du fichier :</span>

<img src="https://imgur.com/BwSXJL1.png" align=center style="width: 400px;">

**Regarder ! Nous avons bel et bien intercepté une communication entre deux siths !
Nous pouvons aperçevoir un message codé que nous allons devoir déchiffrer !**

### <span style="color:red">4) Le Jedi menacé est :</span>

<img src="https://imgur.com/q1ynUMr.png" align=center style="width: 400px;">

**Le Maître Jedi Guillemin est menacé ! Il faut le sauver avant qu'il ne soit trop tard !**

### <span style="color:red">5) Décodons le message des Siths</span>

Le chiffrement utilisée est un chiffrement par substitution (une lettre du message est remplacée par une autre dans le message chiffré)</br>

Heureusement, une personne infiltrée nous a fourni au prix de sa vie, la correspondance de la substitution 
<img src="https://imgur.com/jjKKBCk.png" align=center style="width: 800px;">



In [2]:
from scapy.all import *

#ouverture/création de mon fichier pdf SW2
fichier = open('SW2.pdf', 'r')# r : lire le fichier 

texte = input(str("Entrez la chaîne à traduire :"))

decrypt={'W' : 'A', 
      'S' :'B', 
      'E' : 'C', 
      'R' : 'D' , 
      'D' : 'E' , 
      'X' : 'F' , 
      'C' : 'G', 
      'F' : 'H' , 
      'T' : 'I' , 
      'Y' : 'J' , 
      'G' : 'K', 
      'V' : 'L' , 
      'B' : 'M' , 
      'H' : 'N' , 
      'U' : 'O' , 
      'N' : 'P' , 
      'J' : 'Q' , 
      'I' : 'R' , 
      'O' : 'S' , 
      'K' : 'T' , 
      'L' : 'U' , 
      'M' : 'V' , 
      'P' : 'W' , 
      'A' : 'X' , 
      'Z' : 'Y' , 
      'Q' : 'Z' ,
      "'" : "'" ,
      " " : " " ,
      "&" : "&" ,
      "." : "." ,
      ":" : ":" }

l= [] #liste vide
for i in texte : #parcours tout le texte
    l.append(decrypt[str(i)]) #on ajoute dans notre liste, la clé de décryptage pour chaque lettre 
l = "".join(l) #permet de réunir toutes les "valeur de la liste dans une valeur"
print("\n")
print(l) #affiche la liste

print("\n##########################")
print("Récupération terminé !")
print("##########################")

Entrez la chaîne à traduire :RWIG OTRTULO HULO RDMUHO DEIWODI VDO YDRT DK KULKDO VDO NDIOUHHDO JLT VDO  OLNNUIKDHK. TV DOK KDBNO RD XIWNNDI LH CIWHR EULN DK XWTID  EUBNIDHRID W HUO DHHDBTO JL'TVO RUTMDHK OD OULBDKKID W HUKID  NLTOOWHED. IDYUTHO HUKID XVUKKD RWHO VD OZOKDBD RD VW SUIRLID BDRTWHD  RWHO KIUTO YULIO DK BDHD LHD WKKWJLD EUHKID VW NVWHDKD WHRU.  TV HD RUTK NVLO IDOKDI WLELH OLIMTMWHK OLI EDKKD NVWHDKD DK  VW IWED WJLWVTOF RUTK RTONWIWTKID. RD NVLO VD YDRT RUHK YD KD YUTHO VW NFUKU DOK LH YDRT I&K. TV W  RDO JLWVTKDO DAEDNKTUHHDVVDO NULI V'THKDIEDNKTUH RDO  EUBBLHTEWKTUHO. KL RUTO VD KLDI. OUTO OWHO NTKTD. RWIG NVWCDLTO


DARK SIDIOUS NOUS DEVONS ECRASER LES JEDI ET TOUTES LES PERSONNES QUI LES  SUPPORTENT. IL EST TEMPS DE FRAPPER UN GRAND COUP ET FAIRE  COMPRENDRE A NOS ENNEMIS QU'ILS DOIVENT SE SOUMETTRE A NOTRE  PUISSANCE. REJOINS NOTRE FLOTTE DANS LE SYSTEME DE LA BORDURE MEDIANE  DANS TROIS JOURS ET MENE UNE ATTAQUE CONTRE LA PLANETE ANDO.  IL NE DOIT PLUS RESTER AUCUN S

<span style="color:red">**Nous avons bien décrypté le message et le Maître Guillemin est en DANGER !**</span> </br>
Nous lui avons envoyé un **hologram** sur son adresse `willy.guillemin@iut-velizy.uvsq.fr` **à temps** !

<img src="https://imgur.com/37e6P81.png" align=center style="width: 600px;">


# <span style="color:darkgreen">Chapitre 5 : Récupération automatique du numéro de port TCP négocié pour le transfert et du fichier</span>

### <span style="color:red">1) Création pas à pas des commandes pour récupérer le numéro de port négocié pour le transfert de données ainsi que le nom du fichier à transférer</span>

In [69]:
from scapy.all import *

n = 0 #variable n initialisé à 0
# rdpcap lit un fichier pcap ou pcapng (format Wireshark)
paquets=rdpcap("Wireshark/McDiarmid.pcapng")
print("##############################################################")
print("Récupération du numéro de port pour le transfert de données...")
print("##############################################################\n")

#####################################################################
# RÉCUPÉRATION DU PORT NÉGOCIÉ
#####################################################################

for i in range(len(paquets)): # pour i dans compris dans les paquets
    if paquets[6][Raw].load.split(sep=None)[0].decode('latin1') == 'PORT' : #si les paquets[6] sont égaux à 'PORT'
        n = n+1 # ajouter 1 à n à chaque fois que nous passons dans la boucle
        port1 = paquets[6][Raw].load.decode('latin1') # port1 sera le paquet numéro 6 de ma capture en données brute décodé en latin1
        port11 = port1.split(",")[4] # Séparer la 6eme case par des ',' jusq'au 4eme charactère pour obtenir le premier port
        port2 = paquets[6][Raw].load.decode('latin1')
        port22 = port1.split(",")[5]  # Séparer la 6eme case par des ',' jusq'au 5eme charactère pour obtenir le premier port
        portneg = int(port11)*256+int(port22) # Formule pour calculer le port négocier : Port1 * 256 + port2
        # Afficher le numéro de la trame de transfert ainsi que son port
        
print(f"Le port 1 est {port11} et le deuxième est {port22}")
print(f"Le port négocié est : {portneg}")


print("\n#################################")
print("Récupération du nom du fichier...")
print("#################################\n")

print("Le nom du fichier est :", paquets[14][Raw].load.split(sep=None)[1].decode('latin1'))

print("\n##########################")
print("Récupération terminé !")
print("##########################")

##############################################################
Récupération du numéro de port pour le transfert de données...
##############################################################

Le port 1 est 219 et le deuxième est 167

Le port négocié est : 56231

#################################
Récupération du nom du fichier...
#################################

Le nom du fichier est : attaque_sith.pdf

##########################
Récupération terminé !
##########################


- Nous avons bien **récupérer** le numéro de port négocié pour le **transfert** de données (mais en dur).</br>
Le port négocié est **56231**. Cela permet d'éviter les conflits de ports.


- Nous avons bien **récupérer** le nom du fichier mais nous l'avons fais en dur. </br>
Nous voulons créer un programme pour récupérer automatiquement le nom du fichier et le fichier en plus !</br>

Maître Jedi Guillemin a besoin de notre aide pour intercepter les futurs messages des Siths (s'ils sont assez bête pour garder leur méthode de cryptage) !


### <span style="color:red">2) Récupérer automatiquement le numéro de port négocié pour le transfert, le nom du fichier et le fichier</span>

- Nous avons bien ajouté la **récupération** du nom du fichier **automatiquement** et nous avons bien **demandé** si l'utilisateur voulait **récupérer** le fichier en plus.

Désormais, nous devons réaliser l'**automatisation** du port **négocié**.

Toutes les explications sont dans le code sous forme de **commentaires** :

In [74]:
from scapy.all import *

x = 0 # variable utilisé pour trouver le nom du fichier
n = 0 #variable n initialisé à 0
# rdpcap lit un fichier pcap ou pcapng (format Wireshark)
paquets=rdpcap("Wireshark/McDiarmid.pcapng")
print("##############################################################")
print("Récupération du numéro de port pour le transfert de données...")
print("##############################################################\n")
# Choix du port par l'utilisateur
print("##############################################################")
paquet = input(str("Port numéro 20 ou 21 ? "))
print("##############################################################\n")

#####################################################################
# RÉCUPÉRATION DU PORT NEGOCIÉ
#####################################################################
for x in range(len(paquets)): # pour i dans compris dans les paquets
    if paquets[x][Raw].load.split(sep=None)[0].decode('latin1') == 'PORT' : #si les paquets[x] au caractère 0 sont égaux à 'PORT'
        n = n+1 # ajouter 1 à n à chaque fois que nous passons dans la boucle
        port1 = paquets[x][Raw].load.decode('latin1') # port1 sera le paquet numéro x de ma capture en données brute décodé en latin1
        port11 = port1.split(",")[4] # Séparer la 6eme case par des ',' jusq'au 4eme charactère pour obtenir le premier port
        port2 = paquets[x][Raw].load.decode('latin1')
        port22 = port1.split(",")[5]  # Séparer la 6eme case par des ',' jusq'au 5eme charactère pour obtenir le premier port
        portneg = int(port11)*256+int(port22) # Formule pour calculer le port négocier : Port1 * 256 + port2
        # Afficher le numéro de la trame de transfert ainsi que son port
        
print(f"Le port 1 est {port11} et le deuxième est {port22}")
print(f"Le port négocié est : {portneg}")
    
#####################################################################
# RÉCUPÉRATION DU NOM DU FICHIER SI LE PORT EST LE NUMERO 20
#####################################################################

# Si le paquet est du port '20'
if paquet == '20':
    print("#################################")
    print("Récupération du nom du fichier...")
    print("#################################\n")

    for x in range(len(paquets)): # Pour x compris dans les paquets
        if (paquets[x][Raw].load.split(sep=None)[0].decode('latin1')) == 'RETR': #  Si les paquets x au champ désiré 0 sont égaux à la chaine de caractère 'RETR'
            print("Le nom du fichier est :", paquets[x][Raw].load.split(sep=None)[1].decode('latin1')) # alors écrire au champ désiré 1 décodé au format latin1
else:
    print("Il n'y a pas d'envoie de fichier sur le port 21")

#####################################################################
# POUR RÉCUPÉRER LE FICHIER AU FORMAT PDF SI LE PORT EST 20
#####################################################################

# Si le paquet est du port '20'
if paquet == '20':
    # rdpcap lit un fichier pcap ou pcapng (format Wireshark)
    paquets=rdpcap("Wireshark/McDiarmid.pcapng")
    print("\n##############################################################")
    recup = input(str("Souhaitez-vous récupérer le fichier ? (o/n) "))
    print("##############################################################\n")

    if recup == 'o':
        #ouverture/création de mon fichier pdf SW2
        fichier = open('SW2.pdf', 'ab')# a : ajouter du contenu et b : pour le format bytes

        # pour i dans compris dans les paquets
        for i in range(len(paquets)):
            if paquets[i][TCP].sport == 20: #si les paquets[i] de protocole TCP de port source 20
                fichier.write(bytes(paquets[i][Raw]))#alors ça écrit dans 'fichier' en bytes, les paquets du port source 20
        fichier.close() #ferme 'fichier'
        print("Le fichier a été récupérer")
    else:
        print("Le fichier n'a pas été récupéré par décision de l'utilisateur")
#####################################################################

print("\n##########################")
print("Programme terminé !")
print("##########################")

##############################################################
Récupération du numéro de port pour le transfert de données...
##############################################################

##############################################################
Port numéro 20 ou 21 ? 20
##############################################################

Le port 1 est 219 et le deuxième est 172

Le port négocié est : 56236
#################################
Récupération du nom du fichier...
#################################

Le nom du fichier est : attaque_sith.pdf

##############################################################
Souhaitez-vous récupérer le fichier ? (o/n) o
##############################################################

Le fichier a été récupérer

##########################
Programme terminé !
##########################


# <span style="color:darkgreen">Chapitre 6 : Attaque MITM (Man In The Middle)</span>

<div class="alert alert-block alert-info bilan">
Pour éviter que les Jedis me transmettent les fichiers qu'ils subtilisent à l’ennemi.</br>
Il est souhaitable que nous puissions récupérer le login/mot de passe et les fichiers en temps réel lorsque nous interceptons les communications des Siths avec nos attaques MITM
</div>

Tout d'abord, qu'est-ce qu'une attaque **Man In The Middle** ?
<div class="alert alert-block alert-info bilan">

Une attaque **Man In The Middle** est une cyberattaque dans laquelle l'attaquant relaie secrètement et éventuellement modifie les communications entre deux parties qui croient communiquer directement l'une avec l'autre, car l'attaquant s'est inséré entre les deux parties
    
</div>


### <span style="color:red">1) Création d'un script qui :</span>

- a. Sniff les paquets FTP
- b. Récupère le login et mot de passe de l’utilisateur
- c. Identifie le numéro de port négocié pour un transfert de fichier ainsi que le nom du fichier et la fin de transfert
- d. Enregistrer le fichier avec le nom de fichier utilisé pour le transfert

**Qu'est-ce que le sniffing ?**
<div class="alert alert-block alert-info bilan">
Le sniffing est une technique qui consiste à se mettre en écoute sur le réseau afin de voir quels sont les échanges de données. 
</div>
    
Pour cette partie, nous **voulons** sniffer les paquets FTP :

**sniff** permet de capturer le trafic réseau à partir d'une ou plusieurs interfaces. sniff() a 6 options :

- `filter` : filtre les paquets à l'intérieur du noyau Linux ce qui rend le filtrage très rapide.
- `count` : nombre de paquet à capturer.
- `prn` : nom de la fonction à appliquer à chaque paquet reçu.
- `lfilter` : filtre les paquets à l'aide d'une fonction Python, on pourra utiliser une fonction lambda. 
- `timeout` : durée de la capture, par défaut None=∞ (^C pour arrêter)
- `iface` : interface sur laquelle on souhaite capturer les paquets (défaut :all)
- `store` : s'il faut stocker les paquets capturés ou les supprimer (store=0).
- `stopfilter` : fonction à évaluer pour arrêter la capture (la fonction doit retourner true pour arrêter, false pour continuer)

Tout d'abord, nous avons remarqué que le **FTP** était crypté via le chiffrement **TLS**.</br>
<div class="alert alert-block alert-info bilan">
TLS signifie « Transport Layer Security », c’est-à-dire « sécurité de la couche de transport ». 
    
Il permet de **chiffrer** les flux de données sur Internet afin que ces données ne puissent être lues **que** par les **destinataires autorisés**.
</div>

Du coup, nous nous sommes connectés sur le serveur via un mode dit *simple* sans **chiffrement** pour éviter d'avoir des paquets cryptés.

<img src="https://imgur.com/CTojuZ0.png" align=center style="width: 800px;">


### <span style="color:red">a) Interception d'une connexion FTP entre le serveur FTP et le client</span>


In [22]:
from scapy.all import *

def loginpass (paquets):
    try:
        if paquets.haslayer(Raw): # si les paquets ont la couche Raw (données brutes) :
            if paquets[Raw].load.split(sep=None)[0].decode('latin1') == "USER":  # si dans la première valeur de la liste paquet[i][Raw] = à USER :#
                user = paquets[Raw].load.split(sep=None)[1].decode('latin1')  # prends la deuxième valeur de la liste paquet[i][Raw]#
                print("le nom de l'utilisateur est : ", user)
            if paquets[Raw].load.split(sep=None)[0].decode('latin1') == "PASS":  # même chose pour le password#
                password = paquets[Raw].load.split(sep=None)[1].decode('latin1')  # même chose pour le password#
                print("le mot de passe  est : ", password)
    except:
        pass
sniff(prn=loginpass,filter ="tcp port 20 or 21")

<Sniffed: TCP:0 UDP:0 ICMP:0 Other:0>

Voici un aperçu du résultat de notre programme dans *PyCharm*, cela ne fonctionne malheureusement pas dans le Notebook :

<img src="https://imgur.com/p4OLOh7.png" align=center style="width: 800px;">


<img src="https://www.consultancy.nl/profile/media/conclusion-spotlight-2018-05-17-123251893.png" align=center style="width: 400px;">

Nous n'avons malheureusement pas pu finir sachant que notre code parait fonctionnel. </br>
Nous sommes allés chercher de l'aide pour voir si nos camarades padawan avaient réussis et le code paraissait semblable au nôtre, ce qui nous frustre énormément.</br>
Nous avons préférés terminer le notebook, le paufiner pour qu'il soit agréable et intuitif à lire et bien expliqué ce que nous avons compris et fait durant cette SAE.</br>
Nous avons fais le maximum pour détailler et expliquer les programmes, les applications, nos méthodes, nos chapitres, ...</br>



**Notre SAE se résume par le fait que :**
- Nous avons bien découvert l'interception du transfert de fichier entre deux Siths sur WireShark
- Nous avons bien récupèré le login et password d'un Sith
- Nous avons bien installé et configuré FileZilla
- Nous avons bien crée un script permettant de récupérer en dur un login et mot de passe
- Nous avons bien crée un script permettant de récupérer automatiquement un login et un mot de passe
- Nous avons bien crée un programme qui récupère toutes les données transférées dans la capture Wireshark et les stockes dans un fichier nommé SW2.pdf
- Nous avons bien crée un programme qui permet de déchiffrer le message du Dark Sidious.
- Nous avons bien récupèré automatiquement le numéro de port négocié pour le transfert, le nom du fichier et le fichier.
- Nous avons bien récupèré le login et mot de passe de l’utilisateur automatiquement via une connexion FTP

<img src="http://www.pngmart.com/files/7/Check-Background-PNG.png" align=right style="width: 50px;"></img>

**Ce que nous n'avons pas terminés dans le dernier chapitre:**

- a. Sniff les paquets FTP
- c. Identifie le numéro de port négocié pour un transfert de fichier ainsi que le nom du fichier et la fin de transfert
- d. Enregistrer le fichier avec le nom de fichier utilisé pour le transfer

<img src="http://sweetclipart.com/multisite/sweetclipart/files/x_mark_red.png" align=right style="width: 50px;"></img>



*PS : McDiarmid est un acteur mondialement célèbre pour avoir joué le rôle de l'empereur Sheev Palpatine / Dark Sidious dans la saga Star Wars.  Il apparaît dans 6 des épisodes de la saga ainsi que dans la série d'animation Star Wars Rebels.*